File: backend/app/schemas/company.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime
from app.models.company import CompanyStatus


# Shared fields between creating and reading a company.
# Other schemas will inherit from this to avoid repeating fields.
class CompanyBase(BaseModel):
    name: str
    person_in_charge: str
    category: Optional[str] = None
    contact_person: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    status: CompanyStatus = CompanyStatus.NOT_CONTACTED
    offer_amount: float = 0.0
    priority: Optional[str] = None
    notes: Optional[str] = None
    meeting_date: Optional[datetime] = None
    follow_up_date: Optional[datetime] = None


# Used when the frontend sends data to CREATE a new company.
# It only needs what CompanyBase already defines — no id, no timestamps,
# since the database generates those automatically.
class CompanyCreate(CompanyBase):
    pass


# Used when the frontend sends data to UPDATE an existing company.
# Every field is Optional here because an edit might only change
# one field (e.g. just the status) — we don't want to force resending everything.
class CompanyUpdate(BaseModel):
    name: Optional[str] = None
    person_in_charge: Optional[str] = None
    category: Optional[str] = None
    contact_person: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    status: Optional[CompanyStatus] = None
    offer_amount: Optional[float] = None
    priority: Optional[str] = None
    notes: Optional[str] = None
    meeting_date: Optional[datetime] = None
    follow_up_date: Optional[datetime] = None


# Used when sending a company back to the frontend (e.g. in GET requests).
# This includes everything CompanyBase has, PLUS fields the database
# generates itself: id, created_at, updated_at.
class CompanyOut(CompanyBase):
    id: int
    created_at: datetime
    updated_at: datetime

    # This tells Pydantic: "it's okay to build this schema directly from
    # a SQLAlchemy model object, not just from a dictionary."
    # Without this, passing a Company database object to CompanyOut would fail.
    class Config:
        from_attributes = True

File: frontend/types/company.ts
// This mirrors CompanyStatus enum from your backend models/company.py exactly.
// Keeping the string values identical is critical — these are compared
// directly against what the API returns.
export type CompanyStatus =
  | "not_contacted"
  | "negotiation"
  | "meeting_scheduled"
  | "confirmed"
  | "follow_up_required"
  | "rejected";

// Mirrors your CompanyOut Pydantic schema field-for-field.
export interface Company {
  id: number;
  name: string;
  person_in_charge: string;
  category: string | null;
  contact_person: string | null;
  phone: string | null;
  email: string | null;
  status: CompanyStatus;
  offer_amount: number;
  priority: string | null;
  notes: string | null;
  meeting_date: string | null;   // dates come from JSON as ISO strings, not Date objects
  follow_up_date: string | null;
  created_at: string;
  updated_at: string;
}

// Mirrors CompanyCreate — what we send when adding a new company.
export interface CompanyCreate {
  name: string;
  person_in_charge: string;
  category?: string;
  contact_person?: string;
  phone?: string;
  email?: string;
  status?: CompanyStatus;
  offer_amount?: number;
  priority?: string;
  notes?: string;
  meeting_date?: string;
  follow_up_date?: string;
}

// Mirrors CompanyUpdate — every field optional, for partial edits.
export type CompanyUpdate = Partial<CompanyCreate>;